### DISCUSSION TITLE
Get Route/Path name from function to be used for logging

### DISCUSSION THREAD

**Discussion by roshanjrajan** [actix/actix-web#2576] [Q&A] [answered]
Hi! I was wondering if it is possible to extract the path name to be used for logging purposes. I have many of these created but wanted to keep track of what requests are hit the most. Is there a way to get this information? I am looking to get the string "/api/endpoint1" in the example below.

```
#[get("/api/endpoint1")] 
async fn api_request() -> Result<String> {
    Ok(String::from("Hello World"))
}
```

**Comment by robjtede** [comment, marked as answer]
```rust
#[get("/api/endpoint1")] 
async fn api_request(req: HttpRequest) -> Result<String> {
    let path = req.path();
    ...
}
```

For logging, there's a built in Logger middleware (search docs).

**Comment by roshanjrajan** [comment]
Thanks for the response @robjtede. I was able to find ```match_pattern()``` which works better for my use case! But I was hitting this issue where I was trying to get ```match_pattern()``` on ServiceRequest, but would keep getting None. I tried putting the wrap_fn before and after my .service calls (looks like below). I can see this info populated when checking in the HttpRequest handler but wanted to create a generic logging handler in wrap_fn. Is there anything that I can do here?
``` rust
App::new()
.wrap_fn(|req, srv| {
	let path = req.match_pattern() // Returns None
    ...
}
.service(a)
.service(b)
.wrap_fn(|req, srv| {
	let path = req.match_pattern() // Also returns None
    ...
}
```

**Reply by robjtede** [reply]
Firstly, your code sample is equivalent to the following. See that the order of the `service` calls relative to the `wrap_fn` calls does not matter.
```
App::new()
.wrap_fn(|req, srv| {
	let path = req.match_pattern() // Returns None
    ...
}
.wrap_fn(|req, srv| {
	let path = req.match_pattern() // Also returns None
    ...
}
.service(a)
.service(b)
```

However, the reason you're seeing `None` is because there is no routing info at this level in the middleware. Middleware **wrap** the router so the pre-router section of a middleware function has no knowledge of the router itself. If you try to get this info after `srv.call(req)` then you'll see `match_pattern` be populated correctly. (This needs documenting better.)

**Comment by roshanjrajan** [comment]
Thank you for the quick response @robjtede! That makes much more sense!
But I am struggling on 1 thing. I don't fully understand how I can access req after the ```srv.call(req)```? I'm assuming I need to clone it but I can't clone the HttpRequest as it is private, ServiceRequest doesn't implement Clone, and I can't break the object using ```into_parts()``` and reconstruct it using ```from_parts()``` because of the constraints on the counter. Is there another way to get this information? Or do I need to create a middleware similar to the one done for Logger?

**Reply by robjtede** [reply]
https://docs.rs/actix-web/latest/actix_web/dev/struct.ServiceResponse.html#method.request

**Comment by roshanjrajan** [comment]
Exactly what I am looking for! Thank you so much for your help!
